Add priority option for repo configuration

kellyma2 opened the pull request:
Currently there's no user-specified ordering for the defined repositories and because of the way the code works it will (counter intuitively) choose RPMs from repos later in the list.

This change introduces an explicit repo priority so that users can force ordering of repos to choose RPMs from.

kellyma2 committed:
Add priority option for repo configuration

Currently there's no user-specified ordering for the defined
repositories and because of the way the code works it will (counter
intuitively) choose RPMs from repos later in the list.

This change introduces an explicit repo priority so that users can
force ordering of repos to choose RPMs from.

wade-arista commented on pkg/reducer/reducer.go:
```diff
@@ -102,11 +102,17 @@ func (r *RepoReducer) Resolve(packages []string) (matched []string, involved []*
+				if selected.Repository.Priority > p.Repository.Priority {
+					selected = p
+				}
+
```
remove this blank line?

wade-arista commented on pkg/reducer/reducer.go:
```diff
@@ -102,11 +102,17 @@ func (r *RepoReducer) Resolve(packages []string) (matched []string, involved []*
 		if !found {
```
I think this `RepoReducer.Resolve` unit deserves a unit test.

kellyma2 replied:
While I agree in practice, if we want to add unit tests for this I think we should handle this separately as the change here is (relatively) narrow.

wade-arista replied:
Without a test, it's arguably incomplete. Is the problem that it's hard to add a test to cover this new logic? It doesn't have to cover everything that was here already.

kellyma2 committed:
Remove erroneous blank line

kellyma2 committed:
Seperate RepoReducer from CacheHelper implementation

Separating ourselves from the concrete implementation of the
CacheHelper via an interface makes it easier to write tests against
RepoReducer.

kellyma2 committed:
Remove unused `lang` option for reducer

The lang option never gets used anywhere (and is not even hooked up as
a proper argument!) so we can safely prune it.

kellyma2 committed:
Extract Resolve() function for reducer

All of the uses of the reducer have the exact same sequence.  We can
capture this sequence and then use it as our model for testing the
reducer.

kellyma2 committed:
reducer: pull out loadRepos() from Load()

This is the first step in de-couple the repo data loading from the
reducer itself to enable easier testing of the reducer.

kellyma2 committed:
reducer: Decouple loadRepos() from RepoReducer

In order to completely decouple the loading process from the general
reducer algorithm we need to decouple it from the RepoReducer itself.

This change converts to returning the package list and accepting the
inputs as parameters to enable this decoupling.

kellyma2 committed:
reducer: Factor our separate loader

RepoLoader captures the repository package loading process separately
from the reducer and introduces an interface between them to make
testing the reducer easier.

kellyma2 committed:
Refine loader interface

The loader should also capture the provide-map builder so as to keep
the reducer itself fairly simple.

This change pulls the remainder of the reducer's Load() method into
the loader and refines the interface to the loader to help with this.

kellyma2 committed:
Merge pull request #79 from kellyma2/dead-code

Remove unused RPMReader method

kellyma2 committed:
reducer: Add test for loader

Now that we have the loading code exposed in a way that is a bit more
manageable we can add some tests around it.

We also add a set of helpers in a separate source file which will be
used by tests for the reducer itself.

manuelnaranjo committed:
Make use of cmp.Or instead of complex ifs

Reduce some tech debt by using cmp.Or instead of multi level ifs.

manuelnaranjo committed:
Using slices.[Sort,SortFunc]

Will make use of slices.Sort and slices.SortFunc so we can sort
dictionary keys before we send them to the sat resolver

manuelnaranjo committed:
test: validate if sorting helps

Check if by making the input to sat stable the output is also stable

manuelnaranjo committed:
sat: sort inputs to the sat solver for determinism

Make sure the sat solver output is deterministic and reproducible by
making the input order stable

manuelnaranjo committed:
deps: bumps all golang deps

bumping all golang deps to close a few dependabots issues

manuelnaranjo committed:
Merge pull request #87 from rmohr/mnaranjo/bump-go-dependencies

deps: bumps all golang deps

manuelnaranjo committed:
Merge pull request #86 from rmohr/mnaranjo/make-sat-reproducible

Make sat reproducible

kellyma2 committed:
Merge branch 'main' into HEAD

kellyma2 committed:
Merge pull request #85 from kellyma2/reducer-refactor

Refactor reducer to enable testing

kellyma2 committed:
reducer: Add tests for reducer process

This change approximates the reducer operations and provides tests for
them.  It leverages the existing helpers produced for the loader_test
to reduce the repetitiveness of constructing the test inputs.

kellyma2 committed:
Fix broken behaviour and add some tests

The first iteration of this change produced a scenario where we'd drop
packages that have a different version which we previously were
retaining.  This change fixes that and includes differently versioned
packages while accounting for the repository priority otherwise.

kellyma2 committed:
Merge branch 'reducer-test' into priority
